crates/red_knot_python_semantic/src/types/call/arguments.rs

```diff
@@ -1,88 +1,179 @@
+use std::cell::Cell;
+use std::rc::Rc;
+
+use bitflags::bitflags;
+
 use super::Type;
 
-/// Typed arguments for a single call, in source order.
+/// Typed arguments for a single call, in source order, with an optional bound self/cls parameter.
 #[derive(Clone, Debug, Default)]
-pub(crate) struct CallArguments<'a, 'db>(Vec<Argument<'a, 'db>>);
+pub(crate) struct CallArguments<'a, 'db> {
+    bound_self: Option<Argument<'a, 'db>>,
+    arguments: Rc<[Argument<'a, 'db>]>,
+}
 
 impl<'a, 'db> CallArguments<'a, 'db> {
     /// Create a [`CallArguments`] with no arguments.
     pub(crate) fn none() -> Self {
-        Self(Vec::new())
+        Self {
+            bound_self: None,
+            arguments: vec![].into(),
+        }
     }
 
     /// Create a [`CallArguments`] from an iterator over non-variadic positional argument types.
     pub(crate) fn positional(positional_tys: impl IntoIterator<Item = Type<'db>>) -> Self {
         positional_tys
             .into_iter()
-            .map(Argument::Positional)
+            .map(|ty| Argument::positional().with_argument_type(ty))
             .collect()
     }
 
     /// Prepend an extra positional argument.
     pub(crate) fn with_self(&self, self_ty: Type<'db>) -> Self {
-        let mut arguments = Vec::with_capacity(self.0.len() + 1);
-        arguments.push(Argument::Synthetic(self_ty));
-        arguments.extend_from_slice(&self.0);
-        Self(arguments)
+        assert!(
+            self.bound_self.is_none(),
+            "cannot bind multiple self/cls parameters"
+        );
+        Self {
+            bound_self: Some(Argument::synthetic().with_argument_type(self_ty)),
+            arguments: self.arguments.clone(),
+        }
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.arguments.len() + usize::from(self.bound_self.is_some())
     }
 
     pub(crate) fn iter(&self) -> impl Iterator<Item = &Argument<'a, 'db>> {
-        self.0.iter()
+        self.bound_self.iter().chain(self.arguments.as_ref())
     }
+}
 
-    // TODO this should be eliminated in favor of [`bind_call`]
-    pub(crate) fn first_argument(&self) -> Option<Type<'db>> {
-        self.0.first().map(Argument::ty)
+impl<'a, 'db> FromIterator<Argument<'a, 'db>> for CallArguments<'a, 'db> {
+    fn from_iter<T: IntoIterator<Item = Argument<'a, 'db>>>(iter: T) -> Self {
+        Self {
+            bound_self: None,
+            arguments: iter.into_iter().collect(),
+        }
     }
+}
 
-    // TODO this should be eliminated in favor of [`bind_call`]
-    pub(crate) fn second_argument(&self) -> Option<Type<'db>> {
-        self.0.get(1).map(Argument::ty)
+#[derive(Clone, Debug)]
+pub(crate) struct Argument<'a, 'db> {
+    kind: ArgumentKind<'a>,
+    form: Cell<ArgumentForm>,
+
+    /// The inferred type of this argument. Will be `Type::Unknown` if we haven't inferred a type
+    /// for this argument yet.
+    #[allow(clippy::struct_field_names)]
+    argument_type: Cell<Type<'db>>,
```

Comment:
What were your considerations in using Unknown to represent "no type inferred yet" vs using an `Option<Type>` here? Given that `Unknown` is also a valid inferred type for an argument, the latter seems semantically clearer about whether we've inferred a type for this argument yet. But maybe in practice we don't need that clarity?

Reply:
This is moot with the new represenation, where you have to provide a callback to infer each argument type when constructing a `CallArgumentTypes`.  You now either (a) have all argument types uninferred, in which case you have a `CallArguments`, or (b) have inferred all argument types, in which case you have a `CallArgumentTypes`.